AK/Stream.h

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2020, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <AK/Concepts.h>
+#include <AK/Forward.h>
+#include <AK/Span.h>
+#include <AK/StdLibExtras.h>
+
+namespace AK::Detail {
+
+class Stream {
+public:
+    virtual inline ~Stream()
+    {
+        ASSERT(!error() && !fatal());
+    }
+
+    inline bool error() const { return m_error; }
+    inline bool fatal() const { return m_fatal; }
+
+    inline bool handle_error() { return exchange(m_error, false); }
+
+    inline operator bool() const { return !m_error && !m_fatal; }
+
+protected:
+    mutable bool m_error { false };
+    mutable bool m_fatal { false };
+};
+
+}
+
+namespace AK {
+
+class InputStream : public virtual AK::Detail::Stream {
```

Comment:
Why `virtual` inheritance?

Reply:
I was thinking a bit ahead here, like `OutputStream` and `DuplexStream` classes a bit like the `std::istream`, `std::ostream` and `std::iostream` thing. `DuplexSteam` would inherit from both `InputStream` and `OutputStream` thus inheriting `Detail::Stream` twice, then virtual inheritance would be needed. I can remove this for now.

I actually had something like this ready, but I put it aside because I was simply touching too many things at once.